src/basedtyping/transformer.py

```diff
@@ -0,0 +1,231 @@
+        return node
+
+    def visit_Constant(self, node: ast.Constant) -> ast.AST:
+        value = cast(object, node.value)
+        if isinstance(value, int) or (self.string_literals and isinstance(value, str)):
+            return self._literal(node)
+        return node
+
+    def visit_Tuple(self, node: ast.Tuple) -> ast.AST:
+        node = self.generic_visit(node)
+        if not self._implicit_tuple:
+            return self.subscript(self._typing("Tuple"), cast(ast.expr, node))
+        return node
+
+    def visit_Compare(self, node: ast.Compare) -> ast.AST:
+        if len(node.ops) == 1 and isinstance(node.ops[0], ast.Is):
+            result = self.subscript(
+                self._typing("TypeIs"), cast(ast.expr, self.generic_visit(node.comparators[0]))
+            )
+            return self.generic_visit(result)
+        return self.generic_visit(node)
+
+    def visit_IfExp(self, node: ast.IfExp) -> ast.AST:
+        if (
+            isinstance(node.body, ast.Compare)
+            and len(node.body.comparators) == 1
+            and isinstance(node.body.ops[0], ast.Is)
+        ):
+            node.body = self.subscript(
+                self._typing("TypeGuard"),
+                cast(ast.expr, self.generic_visit(node.body.comparators[0])),
+            )
+        return self.generic_visit(node)
+
+    def visit_FunctionType(self, node: ast.FunctionType) -> ast.AST:
+        node = self.generic_visit(node)
+        assert isinstance(node, ast.FunctionType)
+        return self.subscript(
+            self._typing("Callable"),
+            ast.Tuple([ast.List(node.argtypes, ctx=ast.Load()), node.returns], ctx=ast.Load()),
+        )
+
+    def visit_BinOp(self, node: ast.BinOp) -> ast.AST:
+        node = self.generic_visit(node)
+        if isinstance(node, ast.BinOp) and isinstance(node.op, ast.BitAnd):
+            node = self.subscript(
+                self._basedtyping("Intersection"),
+                ast.Tuple([node.left, node.right], ctx=ast.Load()),
+            )
+        return node
+
+
+def _eval_direct(
+    value: typing.ForwardRef,
+    globalns: dict[str, object] | None = None,
+    localns: dict[str, object] | None = None,
+) -> object:
+    return eval_type_based(value, globalns, localns, string_literals=False)
+
+
+def eval_type_based(
+    value: typing.ForwardRef,
+    globalns: dict[str, object] | None = None,
+    localns: dict[str, object] | None = None,
+    *,
+    string_literals: bool,
+) -> object:
+    """Like `typing._eval_type`, but supports based typing features.
+    Specifically, this transforms `1 | 2` into `typing.Union[Literal[1], Literal[2]]`
+    and `(int) -> str` into `typing.Callable[[int], str]` etc.
+    """
+    tree: ast.AST
+    try:
+        tree = ast.parse(value.__forward_arg__, mode="eval")
+    except SyntaxError:
+        arg = value.__forward_arg__
+        if arg.startswith("def"):
+            arg = arg[3:]
+        tree = ast.parse(arg.lstrip(), mode="func_type")
```

Comment:
what if it's a syntax error caused by something else and the type name happens to start with `def`? eg:
```py
foo: defined_function | () -> int
```

Reply:
ergh

Reply:
i think that's an unsupported scenario, basedmypy doesn't handle it, you have to say:
```py
foo: def_ | "() -> int"
```

Reply:
what that means

Reply:
you always have to put function types in quotes

Reply:
so wouldnt this trigger that handling?
```py
foo: "defined_function | () -> int"
```

Reply:
trueeeee, how to fix it?

Reply:
oh, I know, I'll sort it out